### DISCUSSION TITLE
How does multi-ISA work?

### DISCUSSION THREAD

**Discussion by xiota** [PCSX2/pcsx2#10586] [General]
Most response that most helps to elucidate how multi-ISA works is [this](https://github.com/PCSX2/pcsx2/discussions/10586#discussioncomment-8060103).

The other issue appears to be some parallelism/race condition in `cmake` that I will ask about elsewhere.

-----

What is multi-ISA?  How is it supposed to work?  Should `DISABLE_ADVANCE_SIMD` be renamed?

Notably, by adding a `message` commands before lines 93 and 97:

https://github.com/PCSX2/pcsx2/blob/72787d103f4740c9412d98a52e332e98d4fc7236/cmake/BuildParameters.cmake#L92-L99

```cmake
 if (DISABLE_ADVANCE_SIMD)
 	message("setting individual flags")
 	add_compile_options("-msse" "-msse2" "-msse4.1" "-mfxsr")
 else()
 	# Can't use march=native on Apple Silicon.
 	if(NOT APPLE OR "${CMAKE_HOST_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
 		message("setting march=native")
 		add_compile_options("-march=native")
 	endif()
 endif()
```

When `DISABLE_ADVANCE_SIMD=ON`, `-march=native` appears to be added three times.
```
-- Building for x86-64 (Multi-ISA).
setting march=native
setting individual flags
setting march=native
setting march=native
```

Where/How is `cmake` being told to make multiple passes?  Where/How is it being used?

**Comment by stenzek** [comment]
It's how the SW renderer/IPU/etc have multiple codepaths for SSE4/AVX/AVX2, since they benefit the most from the wider vector width.

I see no reason to rename the variable, it wouldn't really provide any benefit, since it's not something users should be messing with.

The compile flags are not duplicated as far as I can tell.

**Comment by Ziemas** [comment]
The relevant cmake magic is here.
https://github.com/PCSX2/pcsx2/blob/72787d103f4740c9412d98a52e332e98d4fc7236/pcsx2/CMakeLists.txt#L749-L762

**Reply by xiota** [reply]
That explains the multiple passes.  Why doesn't `march=native` cause problems?  Either by allowing illegal instructions to be used or preventing needed instructions from being used when building on old machines.

**Reply by stenzek** [reply]
The CI builds enable multi-isa, and are not built with -march=native. Most files are -msse4.1, with avx/avx2 for the duplicated GS/IPU/etc files.

**Reply by xiota** [reply]
`cmake` appears to be executing `add_compile_options("-march=native")`.

**Reply by stenzek** [reply]
It does not execute that line, nor add `-march=native` if multi-isa is enabled.

**Comment by TellowKrinkle** [comment]
```cmake
if (DISABLE_ADVANCE_SIMD)
	message("setting individual flags")
	add_compile_options("-msse" "-msse2" "-msse4.1" "-mfxsr")
else()
	# Can't use march=native on Apple Silicon.
	if(NOT APPLE OR "${CMAKE_HOST_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
		message("setting march=native")
		add_compile_options("-march=native")
	endif()
endif()
```
The march=native is clearly in the else of the `if (DISABLE_ADVANCE_SIMD)`, if you're hitting it with that set to on, something is very broken.

**Reply by xiota** [reply]
The message is definitely printed, but someone else asserted that the next line is ignored.  That creates new questions, like why is the added message printed?  Why is the `add_compile_options` line ignored?  Why does it happen three times when the earlier `message(STATUS "Building for x86-64 (Multi-ISA).")` is printed only once?  Etc.  (Assertions without explanation, or at least reference to relevant lines of code, aren't helpful.)

**Reply by stenzek** [reply]
It's not printed three times.

Full output:
```
$ cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_PREFIX_PATH=$HOME/deps -DCMAKE_TOOLCHAIN_FILE=$HOME/clang-toolchain.cmake -G Ninja ..
-- The C compiler identification is Clang 16.0.6
-- The CXX compiler identification is Clang 16.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang-16 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++-16 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building for Linux.
-- Building with Clang/LLVM.
-- Building for x86-64.
setting march=native
-- Found Git: /usr/bin/git (found version "2.40.1")
-- Found PCAP: /usr/lib/x86_64-linux-gnu/libpcap.so
-- Performing Test PCAP_LINKS_SOLO
-- Performing Test PCAP_LINKS_SOLO - Success
-- Looking for pcap_get_pfring_id
-- Looking for pcap_get_pfring_id - not found
-- Looking for lzma_auto_decoder in /usr/lib/x86_64-linux-gnu/liblzma.so
-- Looking for lzma_auto_decoder in /usr/lib/x86_64-linux-gnu/liblzma.so - found
-- Looking for lzma_easy_encoder in /usr/lib/x86_64-linux-gnu/liblzma.so
-- Looking for lzma_easy_encoder in /usr/lib/x86_64-linux-gnu/liblzma.so - found
-- Looking for lzma_lzma_preset in /usr/lib/x86_64-linux-gnu/liblzma.so
-- Looking for lzma_lzma_preset in /usr/lib/x86_64-linux-gnu/liblzma.so - found
-- Found LibLZMA: /usr/lib/x86_64-linux-gnu/liblzma.so (found version "5.4.1")
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.13")
-- Found PNG: /usr/lib/x86_64-linux-gnu/libpng.so (found version "1.6.40")
-- Found CURL: /usr/lib/x86_64-linux-gnu/libcurl.so (found version "8.2.1")
-- Could NOT find Vtune (missing: VTUNE_LIBRARIES VTUNE_INCLUDE_DIRS)
-- Found FFMPEG: /usr/include/x86_64-linux-gnu (found version "6.0-6ubuntu1") found components: avcodec avformat avutil swresample swscale
-- Found Zstd: /usr/lib/x86_64-linux-gnu/libzstd.so
-- Found LZ4: /usr/lib/x86_64-linux-gnu/liblz4.so
-- Found the following WebP libraries:
--  WebP (required): /usr/lib/x86_64-linux-gnu/libwebp.so
-- Found WebP: /usr/include
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.8.1")
-- EGL found pkg
-- AIO found
-- LIBUDEV found pkg
-- Found X11: /usr/include
-- Looking for XOpenDisplay in /usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so
-- Looking for XOpenDisplay in /usr/lib/x86_64-linux-gnu/libX11.so;/usr/lib/x86_64-linux-gnu/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- Wayland: Egl requires Client
-- Found Wayland_Client: /usr/lib/x86_64-linux-gnu/libwayland-client.so (found version "1.22.0")
-- Found Wayland_Egl: /usr/lib/x86_64-linux-gnu/libwayland-egl.so (found version "18.1.0")
-- Found Wayland: /usr/lib/x86_64-linux-gnu/libwayland-client.so;/usr/lib/x86_64-linux-gnu/libwayland-egl.so (found version "1.22.0") found components: Egl
-- Found Libbacktrace: /home/user/deps/lib/libbacktrace.a
-- Checking for module 'dbus-1'
--   Found dbus-1, version 1.14.10
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- ryml: using C++ standard: C++20
-- ryml: importing subproject c4core (SUBDIRECTORY)... /home/user/dev/pcsx2/3rdparty/rapidyaml/rapidyaml/ext/c4core
-- c4core: using C++ standard: C++20
-- ryml: -----> target ryml PUBLIC incorporating lib c4core
-- Looking for fts_open
-- Looking for fts_open - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off_t
-- Check size of off_t - done
-- Check size of size_t
-- Check size of size_t - done
-- Performing Test HAVE_FICLONERANGE
-- Performing Test HAVE_FICLONERANGE - Success
-- Performing Test HAVE_NULLABLE
-- Performing Test HAVE_NULLABLE - Success
-- Looking for include file pulse/pulseaudio.h
-- Looking for include file pulse/pulseaudio.h - found
-- Looking for include file alsa/asoundlib.h
-- Looking for include file alsa/asoundlib.h - found
-- Looking for include file jack/jack.h
-- Looking for include file jack/jack.h - not found
-- Looking for include file sndio.h
-- Looking for include file sndio.h - found
-- Looking for include file sys/soundcard.h
-- Looking for include file sys/soundcard.h - found
-- Performing Test HAVE_STDATOMIC
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE
-- Found WrapOpenGL: TRUE
-- Found XKB: /usr/lib/x86_64-linux-gnu/libxkbcommon.so (found suitable version "1.5.0", minimum required is "0.5.0")
-- Found WrapVulkanHeaders: /usr/include
-- Version: 10.1.2
-- Build type: Debug
-- Performing Test HAS_NULLPTR_WARNING
-- Performing Test HAS_NULLPTR_WARNING - Success
Using precompiled headers.
-- Configuring done (3.9s)
-- Generating done (0.1s)
-- Build files have been written to: /home/user/dev/pcsx2/build2
```

**Reply by xiota** [reply]
In your output, you have:
```
-- Building for x86-64.
setting march=native
```
From the following section of code, that indicates you have `DISABLE_ADVANCE_SIMD=OFF`.

https://github.com/PCSX2/pcsx2/blob/72787d103f4740c9412d98a52e332e98d4fc7236/cmake/BuildParameters.cmake#L76-L80

My question specifically states that I have `DISABLE_ADVANCE_SIMD=ON` with the following contained in the output:
```
-- Building for x86-64 (Multi-ISA).
setting march=native
setting individual flags
setting march=native
setting march=native
```
If you do indeed have `DISABLE_ADVANCE_SIMD=ON`, that opens yet more questions, like why you have a different `message` statement printing than I did?

**Reply by stenzek** [reply]
I don't know. Must be something in your environment, or something else you've changed.

```
$ cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_PREFIX_PATH=$HOME/deps -DCMAKE_TOOLCHAIN_FILE=$HOME/clang-toolchain.cmake -G Ninja -DDISABLE_ADVANCE_SIMD=ON  ..
-- The C compiler identification is Clang 16.0.6
-- The CXX compiler identification is Clang 16.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/clang-16 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++-16 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building for Linux.
-- Building with Clang/LLVM.
-- Building for x86-64 (Multi-ISA).
setting individual flags
-- Found Git: /usr/bin/git (found version "2.40.1")
-- Found PCAP: /usr/lib/x86_64-linux-gnu/libpcap.so
```

**Reply by xiota** [reply]
Fresh checkout. Update submodules. Add `message` in front of specified `add_compile_options` in specified file. Run `cmake` (3.28.1) with following options:
```
-DCMAKE_BUILD_TYPE="Release"
-DCMAKE_INSTALL_PREFIX="/usr"
-DDISABLE_BUILD_DATE=ON
-DDISABLE_ADVANCE_SIMD=ON
-DCMAKE_INTERPROCEDURAL_OPTIMIZATION=ON
-DENABLE_TESTS=OFF
-Wno-dev
```

Also tried with `unset AR NM CC CXX CFLAGS CXXFLAGS LDFLAGS MAKEFLAGS`

Output still indicates the `else` block is somehow being executed:
```
-- Building for x86-64 (Multi-ISA).
setting march=native
setting individual flags
setting march=native
setting march=native
```

Without `CC` and `CXX` set, the system defaults to `gcc`/`g++`, and the following is also present:
```
  *************** UNSUPPORTED CONFIGURATION ***************

  You are not compiling PCSX2 with a supported compiler.

  It may not even build successfully.

  PCSX2 only supports the Clang and MSVC compilers.

  No support will be provided, continue at your own risk.

  *********************************************************
```

Build is successful, but much slower than with `clang`.  With `-DCMAKE_VERBOSE_MAKEFILE=ON` added to cmake, looks like `march=native` is not used.  But still strange that the corresponding `message` is printed.

**Reply by stenzek** [reply]
I still get the same output. Also,

1. `CMAKE_INSTALL_PREFIX` has no effect on PCSX2.
2. You shouldn't be overriding CC/CXX, you should use a toolchain file like our CI.

This sounds like it's for a third-party package, in which case, sorry, but you're on your own. We can only support the configurations we use on our CI.

**Reply by xiota** [reply]
> You shouldn't be overriding CC/CXX

I said "also tried" unsetting CC/CXX.  This was to rule out custom configuration issues.

> `CMAKE_INSTALL_PREFIX` has no effect on PCSX2.

Will test that next.

> We can only support the configurations we use on our CI.

That's fine.  Would still like the behavior of the `cmake` scripts and multi-isa explained better.  The variable name `DISABLE_ADVANCE_SIMD` makes it look like the intent is to disable AVX when it seems to do the opposite.

**Reply by stenzek** [reply]
> I said "also tried" unsetting CC/CXX. This was to rule out custom configuration issues.

You're still not using a toolchain file, which is how we build PCSX2 on Linux. GCC, or the "default" compiler is not supported due to what appears to be a compiler bug.

> That's fine. Would still like the behavior of the cmake scripts and multi-isa explained better. 

Users don't mess with CMake, so there's no reason to explain it. If you want to contribute documentation for the CMake options, the relevant page is here: https://github.com/PCSX2/pcsx2/wiki/10-Building-on-Linux

> The variable name DISABLE_ADVANCE_SIMD makes it look like the intent is to disable AVX when it seems to do the opposite.

It prevents -march=native being used, so it's disabling ISA higher than SSE4. Makes sense to me.

**Reply by xiota** [reply]
> It prevents -march=native being used, so it's disabling ISA higher than SSE4. Makes sense to me.

Disabling ISA higher than SSE4 should block use of "multiple codepaths for SSE4/AVX/AVX2".  But from this block, it appears to be intended to enable avx, via `march=haswell` or `march=sandybridge`.

https://github.com/PCSX2/pcsx2/blob/01842a3c6bb459965658d0ba9301fa5467b274ad/tests/ctest/core/CMakeLists.txt#L15-L28

> You're still not using a toolchain file, which is how we build PCSX2 on Linux.

I generated a toolchain file with https://github.com/PCSX2/pcsx2/blob/master/.github/workflows/scripts/linux/generate-cmake-qt.sh.  Output is the same.  Commands in that `else` block appear to be run.  `cmake` options are basically the same as what I used before, except it uses `TRUE` instead of `ON` and sets some options that are already set with env variables or default. 

> GCC, or the "default" compiler is not supported due to what appears to be a compiler bug.

The same behavior is seen regardless of whether `clang` or `gcc` is used.  Problem seems to be caused by something directly related to `cmake`, not the compiler.  The messages are printed by `cmake`.
